Userland/Libraries/LibWeb/XHR/FormData.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2023, Kenneth Myhra <kennethmyhra@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/TypeCasts.h>
+#include <LibJS/Runtime/Completion.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/FileAPI/Blob.h>
+#include <LibWeb/FileAPI/File.h>
+#include <LibWeb/HTML/FormControlInfrastructure.h>
+#include <LibWeb/WebIDL/DOMException.h>
+#include <LibWeb/XHR/FormData.h>
+
+namespace Web::XHR {
+
+// https://xhr.spec.whatwg.org/#dom-formdata
+WebIDL::ExceptionOr<JS::NonnullGCPtr<FormData>> FormData::construct_impl(JS::Realm& realm, Optional<JS::NonnullGCPtr<HTML::HTMLFormElement>>)
+{
+    // FIXME: 1. If form is given, then:
+    // FIXME:    1. Let list be the result of constructing the entry list for form.
+    // FIXME:    2. If list is null, then throw an "InvalidStateError" DOMException.
+    // FIXME:    3. Set this’s entry list to list.
+
+    return construct_impl(realm);
+}
+
+WebIDL::ExceptionOr<JS::NonnullGCPtr<FormData>> FormData::construct_impl(JS::Realm& realm, HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>> entry_list)
+{
+    return MUST_OR_THROW_OOM(realm.heap().allocate<FormData>(realm, realm, move(entry_list)));
+}
+
+FormData::FormData(JS::Realm& realm, HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>> entry_list)
+    : PlatformObject(realm)
+    , m_entry_list(move(entry_list))
+{
+    set_prototype(&Bindings::ensure_web_prototype<Bindings::FormDataPrototype>(realm, "FormData"));
+}
+
+FormData::~FormData() = default;
+
+void FormData::visit_edges(Cell::Visitor& visitor)
+{
+    Base::visit_edges(visitor);
+    for (auto entry : m_entry_list) {
+        for (auto value : entry.value) {
```

Comment:
These should probably be `const&` to not copy these values (and entries).

This might work because there `JS::Handle`s but I think if there handles you don't actually have to visit them right?